src/server/transaction.cc

```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
 // BLPOP where a data must be read from multiple shards before performing another hop.
 OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
   DCHECK(!cb_);
-
   cb_ = std::move(cb);
 
-  // single hop -> concluding.
-  coordinator_state_ |= (COORD_EXEC | COORD_EXEC_CONCLUDING);
+  DCHECK(multi_ || (coordinator_state_ & COORD_SCHED) == 0);   // Only multi schedule in advance.
+  coordinator_state_ |= (COORD_EXEC | COORD_EXEC_CONCLUDING);  // Single hop means we conclude.
 
-  if (!multi_) {  // for non-multi transactions we schedule exactly once.
-    DCHECK_EQ(0, coordinator_state_ & COORD_SCHED);
-  }
-
-  bool schedule_fast = (unique_shard_cnt_ == 1) && !IsGlobal() && !multi_;
-  bool run_eager = false;
+  bool was_ooo = false;
 
-  if (schedule_fast) {  // Single shard (local) optimization.
-    // We never resize shard_data because that would affect MULTI transaction correctness.
+  // If we run only on one shard and conclude, we can avoid scheduling at all
+  // and dispatch a single task to the its shard set.
```

Comment:
`and dispatch a single task to the its shard set` -> `and directly dispatch the task to its destination shard`